engine/utils/collections/StackArray.h

```diff
@@ -453,11 +491,87 @@ class MSArray
      * no garbage data is accessed
      * @return a MIter to iterate over the array
      */
-    inline Utils::MIter<const MSArray<T, S>, T> CreateFIterator() const
+    inline Utils::CIter<MSArray<T, S>, T> CreateFIterator() const
     {
         return {this};
     }
 
+    template<typename F>
+    inline void ForEachI(F&& func)
+    {
+        size_t i = 0;
+        for (auto it = CreateFIterator(); it; ++it)
+        {
+            func(*it, i++);
+        }
+    }
```

Comment:
Why do we need these? Also if we were to do this `typename F` is unsafe. Should be `std::function<void(F*, int32_t)>` or whatever. Loses value of using iterator by dereferencing every single time. I think `i` can be replaced with `it - startIt` if that's available somewhere.

Reply:
Regarding the typename - I've added a template trait guard which specifies that the type must be a function of some sort. This should stop any other type being added. 

These are just helper functions which make iterating over these arrays a little more functional. I personally like being able to string together simple operations using functions but I can understand if this isn't everyone's cup of tea. 

Regarding the `it` comment - I'm not sure what you mean there.